keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,129 @@
+---
+title: Service Account Token for CSI Driver
+authors:
+  - "@zshihang"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-auth
+  - sig-storage
+reviewers:
+  - "@msau42"
+approvers:
+creation-date: 2020-06-09
+last-updated: 2020-06-09
+status: implementable
+---
+
+# Service Account Token for CSI Driver
+
+## Table of Contents
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [User stories](#user-stories)
+- [Proposal](#proposal)
+  - [CSIDriver Object](#csidriver-object)
+  - [Example Workflow](#example-workflow)
+- [Alternatives](#alternatives)
+- [Implementation History](#implementation-history)
+
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a way to obtain service account token for pods that the CSI
+drivers are provisioning volumes for. Since these tokens are valid only for a
+limited period, this KEP will also give the CSI drivers an option to remount
+volumes.
+
+## Motivation
+
+Currently, the only way that CSI drivers acquire service account tokens is to
+directly read the token in the file system. However, this approach has uncharming
+traits:
+
+1.  It will not work for csi drivers which run as a different non-root user than
+    the pods. See
+    [file permission section for service account token](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20180515-svcacct-token-volumes.md#file-permission).
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  neccessary get tokens after the inital mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange for
+  AWS credentials.
+
+## Proposal
+
+Kubelet should inject audience-bounded service account tokens requested by the
+CSI driver from the `TokenRequest` API in `VolumeContext` to
+`NodePublishVolume`.
+
+### CSIDriver Object
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountTokenAudience *string
+}
+```
+
+`ServiceAccountTokenAudience` will be used to set on the created token, which
+will be validated to not be empty string (i.e. kube apiserver).
+`RequiresRemount` is optional, which should be only set when the mounted volumes
+by the CRI driver have TTL and require re-validation on the token.
```

Comment:
for vault, the token is used to authenticate as the pod's service account so that the driver can fetch the secrets stored in vault. however, the secrets have TTL, so it will authenticate again to fetch the newer secrets. in this case, `RequiresRemount==true`.

for other csi that doesn't need re authentication, `RequiresRemount==false`, e.g. cert manager csi driver wants to create CertificateRequest on behalf of the pod the credential of which is only needed in the first mount if the certificate doesn't expire.

Reply:
Does this mean remounting the volume set up by the driver that consumes the service account token? In my driver that refreshes its service account tokens, I don't remount the volume; I use (a port of) [`AtomicWriter`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/atomic_writer.go#L40-L62) to replace the files holding the secrets fetched from Vault.

With secrets (or X.509 certificates) acquired from Vault, having authenticated to Vault with a Kubernetes service account token, there are three things that can expire:
- The secret data  
  (By way of an optional advisory TTL.)
- The Vault token  
  (Usually you can refresh it, but when that offer runs out, you need to authenticate again.)
- The Kubernetes service account token  
  (Minimum validity period from the _TokenRequest_ API is ten minutes.)

Even when the secret data changes, though, it's not necessary to unmount and remount volumes. The CSI driver can replace that data on disk "atomically".

If I've misunderstood what is being remounted here, feel free to ignore my critique.

Reply:
> Does this mean remounting the volume set up by the driver that consumes the service account token?

the remount is triggered by kubelet, which means a sequence of `NodePublishVolume` will be called. this way 
can assure the token not going stale and move the common token fetching logic from different csi drivers to kubelet. 

> I use (a port of) AtomicWriter to replace the files holding the secrets fetched from Vault.

this is good. many existing volumes which requires remount (downwardAPI, projected service account token) are also using AtomicWriter. 

it is true that there is extra overhead incurred by the periodic remount operation in this approach, but you can still use AtomicWriter right? what the difference in overhead between the writes via AtomicWriter and mounts for the secrets store csi driver? the difference in overhead seems minimal for other in tree volumes using AtomicWriter.

Reply:
There's no remounting of volumes necessary, though. The process running in the container may be reading the files that would be unmounted. It introduces a gap during which the files would not be available any longer, and in most cases slightly stale files are better than no files.

For most drivers, unmounting involves recursively deleting files from a directory, and mounting requires creating directories. There may be a mixture of files within such a mounted volume, with differing expiration times. Sweeping it all away and writing it again a short time later is not what these other drivers are doing with `AtomicWriter`, and it's not how my driver works. `AtomicWriter` helps facilitate in-place updates of a tree of files. Deleting all the files and writing them again later is not an atomic update.

Reply:
> There's no remounting of volumes necessary, though.

what if the secrets from vault expired and re authentication is required? i assume the token fetching is done in kubelet as it is not ideal from a security point of view to grant TokenRequest permission to the service account of csi drivers

the remount here means mount calls will be re-executed and no unmount is involved in this logic. that's why the atomical updates is possible.

Reply:
If there's no intervening unmounting, then this sounds less threatening. I now understand that you're describing the driver responding to repeated requests to mount, as opposed to mounting, unmounting, then mounting again. Do we have other examples of drivers being told to do that, using this kind of like sending SIGHUP to ask a process to read its configuration again?

I'm wondering who gets to specify details for the service account tokens like expiration duration and the set of audiences.

Reply:
maybe cert manager driver needs remount if certificates can expire. the existing volumes using this mechanism are downwardAPI and projected volumes.

the cluster admin who creates "CSIDriver" will define the audience. why a set of audiences? i am thinking one csi driver can declare at most 1 audience. i prefer to set the same expiration duration as the pod's service account token but open to add a configuration for it.

Reply:
The _TokenRequest_ API accepts [a set of audiences](https://github.com/kubernetes/api/blob/02edbc1d22f3f0842d77b5cac047f25f8da1b6f9/authentication/v1/types.go#L143-L149). There are various reasons to expect that a server might accept different audiences. We might need to use the same token for more than one consumer, each of them making separate calls to the _TokenReview_ API.

As for expiration, the pod's normal service account token has no expiration duration. Tokens used by a driver should have tighter expiration windows. The operator has to trust that the caller of the _TokenRequest_ API (the kubelet, in your proposal) will pass on this configured duration.

Reply:
I think the term "remount" may be misleading. We don't actually unmount and mount directories, we change out the symlinks to files. We wouldn't be able to actually a new mount without restarting the container because the container already has the directory bind mounted and that doesn't change.

So this "remount" behavior may be limited to only plugins that actually don't use mounts, and only populate contents of directories. @jsafrane I wonder if we need to clarify this in the csi spec. I believe right now the csi spec requires that non-block volumes are mount points, but the reality is a lot of these ephemeral volume types are just directories.

Reply:
take secrets store csi driver as an example, the consumer you mentioned here are secrets store provider? this user cased makes sense to me.

i was referring to the pod's projected service account token that we are migrated to, which has expiry. https://github.com/kubernetes/kubernetes/issues/70679. if the BoundServiceAccountTokenVolume is not turned on, the pod's secret based service account token doesn't have expiry and in this case, we can use the default timeout in TokenRequest.

Reply:
I don't think we should take the lifetime of any of the projected service account tokens in use by a pod into account here. Those tokens can have different audiences and different purposes.

Why would we force users to accept the default validity period for the a _TokenRequest_ API? We're designing a new API here, with the opportunity to include such an optional parameter.

Reply:
fair enough, i can add the option.

Reply:
IMO, term "remount" comes from [Kubernetes RequiresRemount handling](https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L362)

It has never remounted anything, it refreshes atomic writer volumes. Refactoring to "Refresh" would be better, but orthogonal to this enhancement.

Reply:
@jsafrane do you think we need to clarify in the csi spec this new refresh behavior or leave this as a Kubernetes-specific behavior? Although already, the drivers that support this would have to be Kubernetes-specific anyway since they need to make the TokenRequest API call. Right now I don't think it's expected the NodePublishVolume could get called repeatedly.

Reply:
the csi driver would not make api calls. TokenRequest will be called by kubelet and TokenReview is called by storage provider.

Reply:
Still, NodePublish will be called frequently. So far, NodePublish was called only once when mounting the volume for the first time. As long as the NodePublish is called in ephemeral CSI volumes (secrets and similar stuff), they're Kubernetes extension anyway and we don't need to update CSI spec.

AWS EFS does not seem to be interested in periodic NodePublish when the token expires, right @micahhausler?